File: src/skaters/spec.py
"""Symbolic specification for skater pipelines.

A spec is a plain dict that fully describes how to build a skater.
It can be serialized to JSON, hashed, compared, and materialized
into a live skater via `build(spec)`.

Every spec has an "op" key. The grammar:

    spec := {"op": "ema", "alpha": float, "k": int}
          | {"op": "ensemble", "k": int, "skaters": [spec, ...]}
          | {"op": "conjugate", "skater": spec, "transform": transform_spec}
          | {"op": "envelope", "skater": spec, "k": int, "decay": float|null}
          | {"op": "calibrated", "skater": spec, "k": int, "target": float}

    transform_spec := {"op": "diff"}
                    | {"op": "frac", "d": float, "window": int}
                    | {"op": "std", "alpha": float}

Canonical names are derived from specs deterministically:

    ema(0.1)
    diff|ema(0.1)                        # conjugation: transform|skater
    std(0.05)|diff|ema(0.1)              # chains read left-to-right
    ensemble(ema(0.01),ema(0.1),ema(0.3))
    diff|ensemble(ema(0.01),ema(0.1))
    calibrated[diff|ensemble(ema(0.01),ema(0.1))]
    envelope[ema(0.1),decay=0.95]
"""

from __future__ import annotations
import json

from skaters.ema import ema
from skaters.ensemble import precision_weighted_ensemble
from skaters.conjugate import conjugate
from skaters.transform import difference, fractional_difference, standardize
from skaters.envelope import envelope
from skaters.calibrated import calibrated_envelope


# ---------------------------------------------------------------------------
# Build: spec -> live skater
# ---------------------------------------------------------------------------

def build(spec: dict):
    """Materialize a spec dict into a live skater callable.

    The returned skater's __name__ is set to the canonical name.
    """
    op = spec["op"]

    if op == "ema":
        f = ema(alpha=spec["alpha"], k=spec["k"])

    elif op == "ensemble":
        k = spec["k"]
        subs = [build(s) for s in spec["skaters"]]
        f = precision_weighted_ensemble(subs, k=k, floor=spec.get("floor", 1e-6))

    elif op == "conjugate":
        inner = build(spec["skater"])
        t = _build_transform(spec["transform"])
        k = spec["skater"]["k"] if "k" in spec["skater"] else _infer_k(spec["skater"])
        f = conjugate(inner, t, k=k)

    elif op == "envelope":
        inner = build(spec["skater"])
        k = spec["k"]
        f = envelope(inner, k=k, decay=spec.get("decay"))

    elif op == "calibrated":
        inner = build(spec["skater"])
        k = spec["k"]
        f = calibrated_envelope(inner, k=k, target=spec.get("target", 0.6827))

    else:
        raise ValueError(f"Unknown op: {op}")

    f.__name__ = name(spec)
    return f


def _build_transform(spec: dict):
    """Build a (forward, inverse_k) transform pair from a spec."""
    op = spec["op"]
    if op == "diff":
        return difference()
    elif op == "frac":
        return fractional_difference(d=spec["d"], window=spec.get("window", 50))
    elif op == "std":
        return standardize(alpha=spec.get("alpha", 0.05))
    else:
        raise ValueError(f"Unknown transform op: {op}")


def _infer_k(spec: dict) -> int:
    """Walk the spec tree to find k."""
    if "k" in spec:
        return spec["k"]
    if "skater" in spec:
        return _infer_k(spec["skater"])
    if "skaters" in spec:
        return _infer_k(spec["skaters"][0])
    raise ValueError(f"Cannot infer k from spec: {spec}")


# ---------------------------------------------------------------------------
# Name: spec -> canonical string
# ---------------------------------------------------------------------------

def name(spec: dict) -> str:
    """Derive a canonical, deterministic name from a spec."""
    op = spec["op"]

    if op == "ema":
        return f"ema({_fmt(spec['alpha'])})"

    elif op == "ensemble":
        inner = ",".join(name(s) for s in spec["skaters"])
        return f"ensemble({inner})"

    elif op == "conjugate":
        t = _transform_name(spec["transform"])
        s = name(spec["skater"])
        return f"{t}|{s}"

    elif op == "envelope":
        s = name(spec["skater"])
        decay = spec.get("decay")
        if decay is not None:
            return f"envelope[{s},decay={_fmt(decay)}]"
        return f"envelope[{s}]"

    elif op == "calibrated":
        s = name(spec["skater"])
        target = spec.get("target", 0.6827)
        if abs(target - 0.6827) < 1e-4:
            return f"calibrated[{s}]"
        return f"calibrated[{s},target={_fmt(target)}]"

    else:
        raise ValueError(f"Unknown op: {op}")


def _transform_name(spec: dict) -> str:
    op = spec["op"]
    if op == "diff":
        return "diff"
    elif op == "frac":
        w = spec.get("window", 50)
        if w == 50:
            return f"frac({_fmt(spec['d'])})"
        return f"frac({_fmt(spec['d'])},w={w})"
    elif op == "std":
        return f"std({_fmt(spec.get('alpha', 0.05))})"
    else:
        raise ValueError(f"Unknown transform op: {op}")


def _fmt(x: float) -> str:
    """Format a float compactly."""
    if x == int(x):
        return str(int(x))
    s = f"{x:.6g}"
    return s


# ---------------------------------------------------------------------------
# Parse: canonical name -> spec (roundtrip)
# ---------------------------------------------------------------------------

def to_json(spec: dict) -> str:
    """Serialize a spec to JSON."""
    return json.dumps(spec, separators=(",", ":"))


def from_json(s: str) -> dict:
    """Deserialize a spec from JSON."""
    return json.loads(s)


# ---------------------------------------------------------------------------
# Spec constructors (convenience)
# ---------------------------------------------------------------------------

def ema_spec(alpha: float = 0.05, k: int = 1) -> dict:
    return {"op": "ema", "alpha": alpha, "k": k}


def ensemble_spec(*skater_specs: dict, k: int = 1) -> dict:
    return {"op": "ensemble", "k": k, "skaters": list(skater_specs)}


def conjugate_spec(skater_spec: dict, transform_spec: dict) -> dict:
    return {"op": "conjugate", "skater": skater_spec, "transform": transform_spec}


def envelope_spec(skater_spec: dict, k: int = 1, decay: float | None = None) -> dict:
    return {"op": "envelope", "skater": skater_spec, "k": k, "decay": decay}


def calibrated_spec(skater_spec: dict, k: int = 1, target: float = 0.6827) -> dict:
    return {"op": "calibrated", "skater": skater_spec, "k": k, "target": target}


def diff_spec() -> dict:
    return {"op": "diff"}


def frac_spec(d: float = 0.4, window: int = 50) -> dict:
    return {"op": "frac", "d": d, "window": window}


def std_spec(alpha: float = 0.05) -> dict:
    return {"op": "std", "alpha": alpha}

File: rust/src/spec.rs
//! Symbolic specification for skater pipelines. Port of src/skaters/spec.py.
//!
//! A spec is plain data that fully describes how to build a skater. It
//! serializes to the same JSON shape as the Python dicts (an "op" tag),
//! hashes, compares, and materializes into a live skater via `build`.
//!
//! Grammar:
//!
//! ```text
//! spec := leaf(k) | ema(alpha, k) | ensemble(k, skaters) | conjugate(skater, transform)
//! transform_spec := diff | frac(d, window) | std(alpha) | ema_t(alpha)
//! ```

use crate::leaf::Leaf;
use crate::skater::{conjugate, ema, precision_weighted_ensemble, Sk};
use crate::transform::{difference, ema_transform, fractional_difference, standardize, Transform};
use serde::{Deserialize, Serialize};

fn default_window() -> usize {
    50
}

fn default_std_alpha() -> f64 {
    0.05
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(tag = "op")]
pub enum Spec {
    #[serde(rename = "leaf")]
    Leaf { k: usize },
    #[serde(rename = "ema")]
    Ema { alpha: f64, k: usize },
    #[serde(rename = "ensemble")]
    Ensemble {
        k: usize,
        skaters: Vec<Spec>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        floor: Option<f64>,
    },
    #[serde(rename = "conjugate")]
    Conjugate {
        skater: Box<Spec>,
        transform: TransformSpec,
    },
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(tag = "op")]
pub enum TransformSpec {
    #[serde(rename = "diff")]
    Diff,
    #[serde(rename = "frac")]
    Frac {
        d: f64,
        #[serde(default = "default_window")]
        window: usize,
    },
    #[serde(rename = "std")]
    Std {
        #[serde(default = "default_std_alpha")]
        alpha: f64,
    },
    #[serde(rename = "ema_t")]
    EmaT { alpha: f64 },
}

// ---------------------------------------------------------------------------
// Build: spec -> live skater
// ---------------------------------------------------------------------------

/// Materialize a spec into a live skater.
pub fn build(spec: &Spec) -> Sk {
    match spec {
        Spec::Leaf { k } => Sk::Leaf(Leaf::new(*k)),
        Spec::Ema { alpha, k } => ema(*alpha, *k),
        Spec::Ensemble { k, skaters, floor } => {
            let subs: Vec<Sk> = skaters.iter().map(build).collect();
            let mut f = precision_weighted_ensemble(subs, *k);
            if let (Sk::Pw(pw), Some(fl)) = (&mut f, floor) {
                pw.floor = *fl;
            }
            f
        }
        Spec::Conjugate { skater, transform } => {
            let inner = build(skater);
            let t = build_transform(transform);
            conjugate(inner, t, infer_k(skater))
        }
    }
}

fn build_transform(spec: &TransformSpec) -> Transform {
    match spec {
        TransformSpec::Diff => difference(),
        TransformSpec::Frac { d, window } => fractional_difference(*d, *window),
        TransformSpec::Std { alpha } => standardize(*alpha, 1e-8),
        TransformSpec::EmaT { alpha } => ema_transform(*alpha),
    }
}

/// Walk the spec tree to find k.
pub fn infer_k(spec: &Spec) -> usize {
    match spec {
        Spec::Leaf { k } | Spec::Ema { k, .. } | Spec::Ensemble { k, .. } => *k,
        Spec::Conjugate { skater, .. } => infer_k(skater),
    }
}

// ---------------------------------------------------------------------------
// Name: spec -> canonical string
// ---------------------------------------------------------------------------

/// Derive a canonical, deterministic name from a spec.
pub fn name(spec: &Spec) -> String {
    match spec {
        Spec::Leaf { .. } => "leaf".to_string(),
        Spec::Ema { alpha, .. } => format!("ema({})", fmt_num(*alpha)),
        Spec::Ensemble { skaters, .. } => {
            let inner: Vec<String> = skaters.iter().map(name).collect();
            format!("ensemble({})", inner.join(","))
        }
        Spec::Conjugate { skater, transform } => {
            format!("{}|{}", transform_name(transform), name(skater))
        }
    }
}

fn transform_name(spec: &TransformSpec) -> String {
    match spec {
        TransformSpec::Diff => "diff".to_string(),
        TransformSpec::Frac { d, window } => {
            if *window == 50 {
                format!("frac({})", fmt_num(*d))
            } else {
                format!("frac({},w={})", fmt_num(*d), window)
            }
        }
        TransformSpec::Std { alpha } => format!("std({})", fmt_num(*alpha)),
        TransformSpec::EmaT { alpha } => format!("ema_t({})", fmt_num(*alpha)),
    }
}

/// Format a float compactly: integers plain, otherwise Python's `%.6g`.
fn fmt_num(x: f64) -> String {
    if x == libm::trunc(x) && x.abs() < 9.0e15 {
        return format!("{}", x as i64);
    }
    fmt_g6(x)
}

/// Python `f"{x:.6g}"`: 6 significant digits, trailing zeros stripped,
/// scientific notation when the decimal exponent is < -4 or >= 6.
fn fmt_g6(x: f64) -> String {
    let sci = format!("{:.5e}", x);
    let (mant, exp) = sci.split_once('e').expect("exponent");
    let exp: i32 = exp.parse().expect("exponent parse");
    if !(-4..6).contains(&exp) {
        let mant = mant.trim_end_matches('0').trim_end_matches('.');
        format!("{}e{}{:02}", mant, if exp < 0 { "-" } else { "+" }, exp.abs())
    } else {
        let prec = (5 - exp).max(0) as usize;
        let fixed = format!("{:.*}", prec, x);
        if fixed.contains('.') {
            fixed.trim_end_matches('0').trim_end_matches('.').to_string()
        } else {
            fixed
        }
    }
}

// ---------------------------------------------------------------------------
// Serialization
// ---------------------------------------------------------------------------

/// Serialize a spec to compact JSON (same separators as Python).
pub fn to_json(spec: &Spec) -> String {
    serde_json::to_string(spec).expect("spec serializes")
}

/// Deserialize a spec from JSON.
pub fn from_json(s: &str) -> Result<Spec, serde_json::Error> {
    serde_json::from_str(s)
}

// ---------------------------------------------------------------------------
// Spec constructors (convenience)
// ---------------------------------------------------------------------------

pub fn leaf_spec(k: usize) -> Spec {
    Spec::Leaf { k }
}

pub fn ema_spec(alpha: f64, k: usize) -> Spec {
    Spec::Ema { alpha, k }
}

pub fn ensemble_spec(skaters: Vec<Spec>, k: usize) -> Spec {
    Spec::Ensemble {
        k,
        skaters,
        floor: None,
    }
}

pub fn conjugate_spec(skater: Spec, transform: TransformSpec) -> Spec {
    Spec::Conjugate {
        skater: Box::new(skater),
        transform,
    }
}

pub fn diff_spec() -> TransformSpec {
    TransformSpec::Diff
}

pub fn frac_spec(d: f64, window: usize) -> TransformSpec {
    TransformSpec::Frac { d, window }
}

pub fn std_spec(alpha: f64) -> TransformSpec {
    TransformSpec::Std { alpha }
}

pub fn ema_t_spec(alpha: f64) -> TransformSpec {
    TransformSpec::EmaT { alpha }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn canonical_names() {
        let spec = conjugate_spec(
            ensemble_spec(vec![ema_spec(0.01, 1), ema_spec(0.1, 1)], 1),
            diff_spec(),
        );
        assert_eq!(name(&spec), "diff|ensemble(ema(0.01),ema(0.1))");
        assert_eq!(name(&ema_spec(0.05, 1)), "ema(0.05)");
        assert_eq!(
            name(&conjugate_spec(leaf_spec(1), frac_spec(0.4, 30))),
            "frac(0.4,w=30)|leaf"
        );
        assert_eq!(
            name(&conjugate_spec(leaf_spec(1), std_spec(0.05))),
            "std(0.05)|leaf"
        );
        // %.6g cases as CPython renders them
        assert_eq!(fmt_g6(0.0000001), "1e-07");
        assert_eq!(fmt_g6(1234560.0), "1.23456e+06");
        assert_eq!(fmt_g6(0.123456789), "0.123457");
    }

    #[test]
    fn json_roundtrip() {
        let spec = conjugate_spec(
            ensemble_spec(vec![ema_spec(0.01, 1), ema_spec(0.1, 1)], 1),
            diff_spec(),
        );
        let s = to_json(&spec);
        assert_eq!(from_json(&s).unwrap(), spec);
        assert!(s.starts_with("{\"op\":\"conjugate\""));
    }
}
